internal/datacoord/segment_reference_manager.go

```diff
@@ -0,0 +1,200 @@
+// Licensed to the LF AI & Data foundation under one
+// or more contributor license agreements. See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership. The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package datacoord
+
+import (
+	"fmt"
+	"path"
+	"strconv"
+	"strings"
+	"sync"
+
+	"go.uber.org/zap"
+
+	etcdkv "github.com/milvus-io/milvus/internal/kv/etcd"
+	"github.com/milvus-io/milvus/internal/log"
+)
+
+type SegmentLock struct {
+	segmentID UniqueID
+	nodeID    UniqueID
+	locKey    string
+}
+
+type SegmentReference struct {
+	etcdKV *etcdkv.EtcdKV
+
+	segmentsLock map[UniqueID][]*SegmentLock
+	lock         sync.RWMutex
+}
+
+func parseLockKey(key string) (segID UniqueID, nodeID UniqueID, err error) {
+	ss := strings.Split(key, "/")
+	// segment lock key consists of at least "/segmentRefer/nodeID/segID"
+	if len(ss) < 4 {
+		return -1, -1, fmt.Errorf("segment lock key is invalid with %s", key)
```

Comment:
This looks like a mixture of C++ and go code.

How about returning, say {0, 0, `err`} at all time when `err != nil` and just log the `err` and the invalid `key` for reference?